File: HisabCore/Sources/HisabCore/SuggestionEngine.swift
import Foundation

/// One transaction as the suggestion engine sees it (the app projects stored
/// rows into this; `effectiveCategory` already accounts for overrides/rules).
public struct SpendRecord: Sendable, Equatable {
    public var merchant: String
    public var amountPaise: Int64
    public var date: Date
    public var direction: Direction
    public var effectiveCategory: String

    public init(merchant: String, amountPaise: Int64, date: Date,
                direction: Direction, effectiveCategory: String) {
        self.merchant = merchant
        self.amountPaise = amountPaise
        self.date = date
        self.direction = direction
        self.effectiveCategory = effectiveCategory
    }
}

public struct RuleSuggestion: Equatable, Sendable {
    public var merchantPattern: String   // normalized cluster key → CategoryRule.pattern
    public var displayMerchant: String   // most common raw merchant in the cluster
    public var totalPaise: Int64
    public var count: Int
}

/// Finds recurring, substantial, uncategorized spending worth one polite
/// prompt. Gates (per the design spec): trailing 90 days; cluster total ≥
/// max(₹500, 2% of the window's total debits); ≥3 transactions across ≥2
/// months; muted merchants never return. Ordered by spend impact.
public enum SuggestionEngine {
    public static func normalize(_ merchant: String) -> String {
        var cleaned = ""
        for ch in merchant.lowercased() {
            if ch.isLetter { cleaned.append(ch) } else { cleaned.append(" ") }
        }
        let tokens = cleaned.split(separator: " ").prefix(3)
        return tokens.joined(separator: " ")
    }

    public static func queue(records: [SpendRecord], now: Date,
                             muted: Set<String>) -> [RuleSuggestion] {
        let windowStart = now.addingTimeInterval(-90 * 86_400)

        var windowDebitTotal: Int64 = 0
        var clusters: [String: [SpendRecord]] = [:]
        for record in records {
            guard record.direction == .debit,
                  record.date >= windowStart, record.date <= now else { continue }
            windowDebitTotal += record.amountPaise
            let uncategorized = record.effectiveCategory == Categorizer.uncategorized
                || record.effectiveCategory == Categorizer.miscellaneous
            guard uncategorized else { continue }
            let key = normalize(record.merchant)
            guard !key.isEmpty, !muted.contains(key) else { continue }
            clusters[key, default: []].append(record)
        }

        let floor: Int64 = 50_000  // ₹500
        let threshold = max(floor, windowDebitTotal * 2 / 100)

        var suggestions: [RuleSuggestion] = []
        for (key, members) in clusters {
            guard members.count >= 3 else { continue }
            let months = Set(members.map { YearMonth(date: $0.date) })
            guard months.count >= 2 else { continue }
            var total: Int64 = 0
            var rawCounts: [String: Int] = [:]
            for member in members {
                total += member.amountPaise
                rawCounts[member.merchant, default: 0] += 1
            }
            guard total >= threshold else { continue }
            let display = rawCounts.max { lhs, rhs in
                lhs.value == rhs.value ? lhs.key > rhs.key : lhs.value < rhs.value
            }?.key ?? key
            suggestions.append(RuleSuggestion(merchantPattern: key,
                                              displayMerchant: display,
                                              totalPaise: total,
                                              count: members.count))
        }
        return suggestions.sorted { lhs, rhs in
            lhs.totalPaise == rhs.totalPaise
                ? lhs.merchantPattern < rhs.merchantPattern
                : lhs.totalPaise > rhs.totalPaise
        }
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/suggestion_engine.dart
/// Rule suggestions. Port of SuggestionEngine.swift — same gates:
/// trailing 90 days; cluster total ≥ max(₹500, 2% of window debits);
/// ≥3 txns across ≥2 months; muted merchants never return.
library;

import 'categories.dart';
import 'domain.dart';
import 'year_month.dart';

class SpendRecord {
  final String merchant;
  final int amountPaise;
  final DateTime date;
  final Direction direction;
  final String effectiveCategory;
  const SpendRecord({
    required this.merchant,
    required this.amountPaise,
    required this.date,
    required this.direction,
    required this.effectiveCategory,
  });
}

class RuleSuggestion {
  final String merchantPattern;
  final String displayMerchant;
  final int totalPaise;
  final int count;
  const RuleSuggestion({
    required this.merchantPattern,
    required this.displayMerchant,
    required this.totalPaise,
    required this.count,
  });
}

class SuggestionEngine {
  static String normalize(String merchant) {
    final cleaned = merchant
        .toLowerCase()
        .split('')
        .map((ch) => RegExp(r'[a-z]').hasMatch(ch) ? ch : ' ')
        .join();
    return cleaned.split(RegExp(r'\s+')).where((t) => t.isNotEmpty).take(3).join(' ');
  }

  static List<RuleSuggestion> queue({
    required List<SpendRecord> records,
    required DateTime now,
    required Set<String> muted,
  }) {
    final windowStart = now.subtract(const Duration(days: 90));

    var windowDebitTotal = 0;
    final clusters = <String, List<SpendRecord>>{};
    for (final record in records) {
      if (record.direction != Direction.debit) continue;
      if (record.date.isBefore(windowStart) || record.date.isAfter(now)) {
        continue;
      }
      windowDebitTotal += record.amountPaise;
      final uncategorized =
          record.effectiveCategory == Categorizer.uncategorized ||
              record.effectiveCategory == Categorizer.miscellaneous;
      if (!uncategorized) continue;
      final key = normalize(record.merchant);
      if (key.isEmpty || muted.contains(key)) continue;
      clusters.putIfAbsent(key, () => []).add(record);
    }

    const floor = 50000; // ₹500
    final threshold = windowDebitTotal * 2 ~/ 100;
    final gate = threshold > floor ? threshold : floor;

    final suggestions = <RuleSuggestion>[];
    clusters.forEach((key, members) {
      if (members.length < 3) return;
      final months = members.map((m) => YearMonth.fromDate(m.date)).toSet();
      if (months.length < 2) return;
      var total = 0;
      final rawCounts = <String, int>{};
      for (final member in members) {
        total += member.amountPaise;
        rawCounts[member.merchant] = (rawCounts[member.merchant] ?? 0) + 1;
      }
      if (total < gate) return;
      var display = key;
      var bestCount = -1;
      rawCounts.forEach((raw, count) {
        if (count > bestCount || (count == bestCount && raw.compareTo(display) < 0)) {
          display = raw;
          bestCount = count;
        }
      });
      suggestions.add(RuleSuggestion(
          merchantPattern: key,
          displayMerchant: display,
          totalPaise: total,
          count: members.length));
    });

    suggestions.sort((l, r) => l.totalPaise == r.totalPaise
        ? l.merchantPattern.compareTo(r.merchantPattern)
        : r.totalPaise.compareTo(l.totalPaise));
    return suggestions;
  }
}
